File: public/codes/2-Regresion/5-Decision_Tree_Regression/decission_tree_regression.py
# -*- coding: utf-8 -*-

# =============================================================================
# Decission Tree Regression
# =============================================================================
# =============================================================================
# --------------------Importando librerias--------------------
# =============================================================================
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================

dataset = pd.read_csv('Position_Salaries.csv')

# Variable independiente:Mayuscula por ser una matriz.
X = dataset.iloc[:,1:2].values

# Variable dependiente:minuscula por ser un vector.
y = dataset.iloc[:,2:3].values

# =============================================================================
# Ajustar la regresion con arboles de decision con el dataset
# =============================================================================

from sklearn.tree import DecisionTreeRegressor
tree_regression=DecisionTreeRegressor(random_state=0)
tree_regression.fit(X,y)

y_pred=tree_regression.predict(X)
# =============================================================================
# Visualizacion de los resultado: Arboles de Decision
# =============================================================================

# Del minimo de la variable independiente en sumandole 0.1 hasta el maximo
X_grid =np.arange(min(X),max(X),0.01)
#
# se ajusta a una matriz
X_grid=X_grid.reshape(len(X_grid),1)

plt.scatter(X,y,color='red')
plt.plot(X_grid,tree_regression.predict(X_grid),color='green') # linea 1
#plt.plot(X,tree_regression.predict(X),color='green') # linea 2
plt.title("Modelo Regresion por arboles de decision")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo en $")
plt.show()

# =============================================================================
# Prediccion de nuestros modelos (Resultados)
# =============================================================================
y_pred=tree_regression.predict([[6.5]])
print(y_pred)

File: public/codes/2-Regresion/5-Decision_Tree_Regression/decission_tree_regression.R
# =============================================================================
# Decission Tree Regression
# =============================================================================

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================
# Estructura de los datos:tipos de empleados y el nivel 
# de cada tipo de empleado y el salario correspondiente.
# Objetivo: asignar un salario correspondientes a un nivel y su posicion dada.
# Filas :10
# Columnas:
#           |Position|Level| (vars independiente)
#           |salary| (var_dependiente)
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

# No se hace ninguna distincion entre variables independientes 
# y variables dependientes en R

# =============================================================================
# Ajustar la regresion con el dataset
# =============================================================================
# crear nuestra variable de regresion aqui
# library(rpart)
tree_regressor=rpart(formula = dataset$Salary ~ dataset$Level,
                     data=dataset,
                     control = rpart.control(minsplit = 1))
summary(tree_regressor)

# =============================================================================
# Visualizacion de los resultado: Modelo {type}
# =============================================================================
# Para mostrar la grafica de datos de entrenamiento

x_grid=seq(min(dataset$Level),max(dataset$Level),0.01)
# Agregando componentes a mostrar
ggplot()+
  # Dobujando los puntos de entrenamiento
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             color="red")+
  # Dobujando la linea de la prediccion, en base al entrenamiento
  geom_line(aes(x=dataset$Level,y=predict(tree_regressor,newdata=dataset)),
  # geom_line(aes(x=x_grid,
  #               y=predict(tree_regressor,
  #                         newdata=data.frame(Level=x_grid))), # ver porque no funciona esta parte----
            color="blue")+
  ggtitle("Prediccion usando regresion con arboles de decision   ")+
  xlab("labelx")+
  ylab("labely")

# =============================================================================
# Prediccion de nuestros modelos (Resultados)
# =============================================================================
y_pred_tree=predict(tree_regressor,newdata=data.frame(Level=6.5))
print(y_pred_tree)

